src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4

```diff
@@ -254,11 +254,12 @@ LITERAL_FALSE:           'false';
 
 CHAR_LITERAL:            '\'' (EscapeSequence | ~['\\\r\n]) '\'';
 
-fragment StringFragment: (EscapeSequence | ~["\\\r\n])*;
+fragment StringFragment: (EscapeSequence | ~["\\\r\n]);
 
-STRING_LITERAL:         '"' StringFragment '"';
+STRING_LITERAL:         '"' StringFragment* '"';
 
-STRING_TEMPLATE_BEGIN:  '"' StringFragment '\\' '{'
+STRING_TEMPLATE_BEGIN:  '"'
+                        { _input.LA(1) != '"' }?
```

Comment:
Extra predicate here that will be helpful in the future when we start parsing text block templates.